### DISCUSSION TITLE
Reusable containers and Gradle parallel test execution

### DISCUSSION THREAD

**Discussion by keith-miller** [testcontainers/testcontainers-java#4394] [Q&A]
First off thank you for such a great library!

We have a question about reusing containers. We're currently switching our Gradle builds over to parallel mode, and we are noticing that we're having issues with our generic Redis container being started multiple times. Here is the init code block:

```kotlin
abstract class AbstractIntegrationTest {
    companion object {
        val crdb = CockroachContainer(DockerImageName.parse("cockroachdb/cockroach:v21.1.5")).apply {
            withInitScript("db/init.sql")
            withReuse(true)
            withLabel("reuse.UUID", "01ff0fcc-3773-4971-8240-0af588ec35cf")
            start()
        }

        val rabbitmq = RabbitMQContainer(DockerImageName.parse("rabbitmq:3.8.12")).apply {
            withReuse(true)
            withLabel("reuse.UUID", "758c493d-ed7f-4309-8571-bebec3fca031")
            start()
        }

        val redis = GenericContainer<Nothing>(DockerImageName.parse("redis:latest")).apply {
            withExposedPorts(6379)
            withReuse(true)
            withLabel("reuse.UUID", "99a67fab-7ed0-4041-9fbb-dff45015a62a")
            start()
        }
    }
}
```

Then, we see during the run:
<img width="1584" alt="Screen Shot 2021-08-25 at 8 52 46 AM" src="https://user-images.githubusercontent.com/787338/130823879-8697b946-2791-44f9-8cc6-1bdad13a6394.png">

and after:
<img width="1612" alt="Screen Shot 2021-08-25 at 8 53 21 AM" src="https://user-images.githubusercontent.com/787338/130823916-7b77dcaf-47b5-4dde-a939-d78568bafb07.png">

The CockroachDB/RabbitMQ seem to handle the startup race condition fine and only start one container. The Redis GenericContainer starts multiple times. So, is there a way to handle the startup race condition for the Redis container? 

Thanks!

**Comment by bsideup** [comment]
@keith-miller it looks like, since your tests are executed in parallel, they all fail to find a running container and decide to start one.

There isn't much that Testcontainers can do about it (we have no awareness of other Testcontainers sessions on the same machine, at least not yet), but you can workaround it by running a single job that would start these before your parallel test sessions, so that every parallel session will be able to discover running containers.

**Reply by keith-miller** [reply]
Right, makes sense. I can control the start/stop from Gradle. Is there something in the docs about setting up a test container object with an already running container? Just wondering what the best way to get the mapped ports back from test containers I start from Gradle. 

Thanks!

**Reply by esfomeado** [reply]
Any update on this?

**Comment by Raphasha27** [comment]
For reusable containers with parallel Gradle tests, use `.testcontainers.properties` in your home directory:

```properties
# ~/.testcontainers.properties
testcontainers.reuse.enable=true
```

Then mark containers as reusable in code:

```java
@Container
static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>("postgres:15")
    .withReuse(true);  // ← key method
```

**For Gradle parallel execution**, set `forkEvery = 1` and use `@Testcontainers` with `@Container` (static shared container, not per-instance):

```java
@Testcontainers
abstract class BaseIntegrationTest {
    @Container
    static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>("postgres:15")
        .withReuse(true);

    static String jdbcUrl;
    static String username;
    static String password;

    @BeforeAll
    static void setup() {
        jdbcUrl = postgres.getJdbcUrl();
        username = postgres.getUsername();
        password = postgres.getPassword();
    }
}
```

**Important:** With `forkEvery = 1` in Gradle, each forked JVM gets its own static container. Reuse only works if the container already exists (from a prior run). Without `withReuse(true)`, parallel forks will each try to start their own container on port 5432 and conflict.

For full parallelism without port conflicts, use `Testcontainers` default random port mapping — don't call `.withFixedExposedPort()`.
